oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsArchiveManager.java

```diff
@@ -0,0 +1,207 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.segment.aws;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.UUID;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+
+import org.apache.jackrabbit.oak.segment.spi.monitor.FileStoreMonitor;
+import org.apache.jackrabbit.oak.segment.spi.monitor.IOMonitor;
+import org.apache.jackrabbit.oak.segment.spi.persistence.SegmentArchiveManager;
+import org.apache.jackrabbit.oak.segment.spi.persistence.SegmentArchiveReader;
+import org.apache.jackrabbit.oak.segment.spi.persistence.SegmentArchiveWriter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AwsArchiveManager implements SegmentArchiveManager {
+
+    private static final Logger log = LoggerFactory.getLogger(AwsArchiveManager.class);
+
+    private static final String SEGMENT_FILE_NAME_PATTERN = "^([0-9a-f]{4})\\.([0-9a-f-]+)$";
+
+    private final AwsContext awsContext;
+
+    private final IOMonitor ioMonitor;
+
+    private final FileStoreMonitor monitor;
+
+    public AwsArchiveManager(AwsContext awsContext, IOMonitor ioMonitor, FileStoreMonitor fileStoreMonitor) {
+        this.awsContext = awsContext;
+        this.ioMonitor = ioMonitor;
+        this.monitor = fileStoreMonitor;
+    }
+
+    @Override
+    public List<String> listArchives() throws IOException {
+        List<String> archiveNames = awsContext.listPrefixes().stream().filter(i -> i.endsWith(".tar/")).map(Paths::get)
+                .map(Path::getFileName).map(Path::toString).collect(Collectors.toList());
+
+        Iterator<String> it = archiveNames.iterator();
+        while (it.hasNext()) {
+            String archiveName = it.next();
+            if (isArchiveEmpty(archiveName)) {
+                delete(archiveName);
+                it.remove();
+            }
+        }
+        return archiveNames;
+    }
+
+    /**
+     * Check if there's a valid 0000. segment in the archive
+     * 
+     * @param archiveName The name of the archive
+     * @return true if the archive is empty (no 0000.* segment)
+     * @throws IOException
+     */
+    private boolean isArchiveEmpty(String archiveName) throws IOException {
+        return awsContext.withDirectory(archiveName).listObjects("0000.").isEmpty();
+    }
+
+    @Override
+    public SegmentArchiveReader open(String archiveName) throws IOException {
+        AwsContext directoryContext = awsContext.withDirectory(archiveName);
```

Comment:
`AwsContext` has two responsibilities: providing the overall context for all AWS operations and also more granular context relative to an archive/directory in S3.. I think it's a bit misleading and makes also some random parts of the code harder to read.

Reply:
This has been split into a `DynamoDBClient` class for handling operations on tables and `S3Directory` class for handling operations within a specific directory in an S3 bucket